dashboard/app/controllers/lti_v1_controller.rb

```diff
@@ -87,30 +98,20 @@ def unauthorized_status
     render(status: :unauthorized, json: {error: 'Unauthorized'})
   end
 
-  def create_and_set_nonce
-    # generate nonce, this will be set as a cookie.
-    nonce = generate_random_string 10
-    # generate hashed nonce, this will be returned to the LTI Platform in the
-    # request params https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes
-    hashed_nonce = Digest::SHA2.hexdigest nonce
-    set_cookie('nonce', nonce)
+  def create_state_and_nonce
+    state = generate_random_string 10
+    nonce = Digest::SHA2.hexdigest(generate_random_string(10))
 
-    hashed_nonce
+    {state: state, nonce: nonce}
   end
 
-  def create_and_set_state
-    state = generate_random_string 10
-    set_cookie('state', state)
-
-    state
+  def write_cache(key, value)
+    CDO.shared_cache.write(key, value.to_json, expires_in: 1.minute)
```

Comment:
Setting an expiration so this will auto-invalidate itself (I hope). I'm not sure if we should be cleaning up cache values after we're done with them. I also need to look into what happens if you attempt to write to an expired cache key pair, like if we generate a state value that matches a previously used but expired state value. This is TODO.

Reply:
There is a [cleanup](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-cleanup) method that removes expired entries. (TODO ask infra) does this run occasionally so we don't have to cleanup entries after we use them?

Reply:
@code-dot-org/infrastructure - There is a [cleanup](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-cleanup) method that removes expired entries. Does this run occasionally so we don't have to cleanup entries after we use them, or should we be deleting entries after use?

Reply:
Answered in slack. Setting an expiration is sufficient for cleaning up entries. Memcache has it's own mechanism for deleting entries when it's memory is full, using a LRU (Least Recently Used) algorithm.

Reply:
Link for reference - https://codedotorg.slack.com/archives/C03CK49G9/p1688142360611859

Reply:
I do just want to confirm here that 1 minute is a sufficiently long expiration time; I'm not entirely sure how these endpoints are used. Does a single action by the user invoke a hit on first the `login` endpoint and then the `authenticate` endpoint, or does the user first take an action to hit `login`, then does other stuff like entering their username and password, then does something to hit `authenticate`?

Reply:
Something else to consider is that it might be wise to namespace the key; since this is the general cache used for everything, something like:

```suggestion
    CDO.shared_cache.write("lti_v1_controller/#{key}", value.to_json, expires_in: 1.minute)
```

Together with a comparable change in the definition of `read_cache`, this could make the feature more resilient.

Reply:
> I do just want to confirm here that 1 minute is a sufficiently long expiration time; I'm not entirely sure how these endpoints are used. Does a single action by the user invoke a hit on first the `login` endpoint and then the `authenticate` endpoint, or does the user first take an action to hit `login`, then does other stuff like entering their username and password, then does something to hit `authenticate`?

Yep, 1 minute is plenty long enough. The `login` endpoint is hit when a user initiates an LTI launch. This kicks off a 3 legged auth flow that happens in a matter of seconds, via redirects between us (the LTI Too) and an LTI Platform (LMS).

Reply:
> Something else to consider is that it might be wise to namespace the key; since this is the general cache used for everything, something like:
> 
> Together with a comparable change in the definition of `read_cache`, this could make the feature more resilient.

Good call, I had read about namespacing keys, and had considered it. I can add it in.

Reply:
Fantastic, thanks!